### PR TITLE
libs/libc/modlib: fix dynamic loader issues

### PR DESCRIPTION
**Pull request by acassis**
## Summary
  * build-globals.sh - Macros for defining symbols etc. based on assembler in use - Use the System.map to get all the nuttx symbols

  * libs/libc/modlib/modlib_globals.S
    - Provide an empty skeleton. If the dynamic loading functions are required then run build-global.sh after a clean build using the skeleton. This will fill out the skeleton with the symbols to be available to dynamically loaded modules.

  * libs/libc/modlib/modlib_loadhdrs.c - Fix case where there are no program headers are avaiable

## Impact
Fix CI
## Testing
build

### CODE REVIEW

File: libs/libc/modlib/modlib_verify.c [hunk 1]
```diff
@@ -28,14 +28,18 @@
...
 
 /****************************************************************************
  * Private Constant Data
  ****************************************************************************/
 
-static const char g_modmagic[EI_MAGIC_SIZE] = EI_MAGIC;
+static const char g_modmagic[EI_MAGIC_SIZE] =
+{
+    0x7f, 'E', 'L', 'F'
+};
```
Comment (pkarashchenko) [inline libs/libc/modlib/modlib_verify.c:42]: I'm not sure why this change is needed as it reverts https://github.com/apache/nuttx/pull/9710

File: libs/libc/modlib/modlib_loadhdrs.c [hunk 2]
```diff
@@ -117,26 +99,43 @@ int modlib_loadhdrs(FAR struct mod_loadinfo_s *loadinfo)
...
+                 (size_t)loadinfo->ehdr.e_phnum;
+      if (loadinfo->ehdr.e_phoff + phdrsize > loadinfo->filelen)
+        {
+          berr("ERROR: Insufficent space for program header table\n");
+          return -ESPIPE;
+        }
+
+      /* Allocate memory to hold a working copy of the program header table */
+
+      loadinfo->phdr = (FAR FAR Elf_Phdr *)lib_malloc(phdrsize);
```
Comment (pkarashchenko) [inline libs/libc/modlib/modlib_loadhdrs.c:116]:
```suggestion
      loadinfo->phdr = (FAR Elf_Phdr *)lib_malloc(phdrsize);
```

File: libs/libc/modlib/modlib_verify.c (hunk 1 above)
Comment (acassis) [reply libs/libc/modlib/modlib_verify.c:42]: Good point, it shouldn't

File: libs/libc/modlib/modlib_loadhdrs.c [hunk 3]
```diff
@@ -88,16 +80,6 @@ int modlib_loadhdrs(FAR struct mod_loadinfo_s *loadinfo)
...
-             (size_t)loadinfo->ehdr.e_phnum;
-  if (loadinfo->ehdr.e_phoff + phdrsize > loadinfo->filelen)
-    {
-      berr("ERROR: Insufficent space in file for program header table\n");
-      return -ESPIPE;
-    }
-
   /* Allocate memory to hold a working copy of the sector header table */
 
   loadinfo->shdr = (FAR FAR Elf_Shdr *)lib_malloc(shdrsize);
```
Comment (pkarashchenko) [inline libs/libc/modlib/modlib_loadhdrs.c:85]:
```suggestion
  loadinfo->shdr = (FAR Elf_Shdr *)lib_malloc(shdrsize);
```

File: build-globals.sh [hunk 4]
```diff
@@ -13,6 +13,8 @@ findEP()
 	SIZE=${#SYM[@]}
```
Comment (pkarashchenko) [inline build-globals.sh:13]: Should this script be places and executed from `tools` folder?

File: libs/libc/modlib/modlib_verify.c [hunk 5]
```diff
@@ -28,6 +28,7 @@
 #include <debug.h>
 #include <errno.h>
 
+#include <nuttx/arch.h>
```
Comment (pkarashchenko) [inline libs/libc/modlib/modlib_verify.c:31]: I think this is also not needed

File: build-globals.sh [hunk 6]
```diff
@@ -80,70 +93,100 @@ EP=(${EPS})
...
 
 for ((i = 0; i < ${#EP[@]}; i++))
 do
-	echo ".L${i}:	.string	\"${EP[$i]}\"" >>${GLOBALS}
+	echo ".l${i}:	.string	\"${EP[$i]}\"" >>${GLOBALS}
 done
 
 cat >>${GLOBALS} <<__EOF__
-	.size	globalNames, . - globalNames
+	SIZE	globalNames
```
Comment (pkarashchenko) [inline build-globals.sh:170]:
```suggestion
	SIZE	globalNames
```

File: build-globals.sh [hunk 7]
```diff
@@ -80,70 +93,100 @@ EP=(${EPS})
...
 #endif
-        .thumb
+	.thumb
 #endif
-        .data
-        .align ALIGN
-	.global globalNames
+	.data
+	.align	ALIGN
+	GLOBAL	globalNames
```
Comment (pkarashchenko) [inline build-globals.sh:159]:
```suggestion
	GLOBAL	globalNames
```

File: libs/libc/modlib/modlib_verify.c [hunk 8]
```diff
@@ -28,6 +28,9 @@
...
+
+#include <nuttx/arch.h>
```
Comment (pkarashchenko) [inline libs/libc/modlib/modlib_verify.c:33]:
```suggestion
```

**Review by pkarashchenko — APPROVED** [review]
LGTM

**Comment by tmedicci** [comment]
Related to issue #9845: 

Is it expected to run `./build-globals.sh` after a clean build and rebuild the firmware to include the symbols at `libs/libc/modlib/modlib_globals.S`, right? Is there any other procedure to perform?
